---
title: "dino_popgen_Fig2_JBW"
output: pdf_document
---

```{r warning=FALSE, message=FALSE, results = "hide}

# load the packages
library(reshape2)
library(pcadapt)
library(cowplot)
library(ggplot2)
library(ggpubr)
library(vcfR)
```

```{r set wd and load data}
#setwd("Desktop/Macbook/WHOI/Classes/FALL_21/final-project")

# import the files
poptab <- read.csv("supp_table_ordered.csv")
path <- "allhqSNPmm80.recode.vcf"
vcf <- read.pcadapt(path, type = "vcf", type.out = "matrix", ploidy = 1)
```

```{r eval = FALSE, results="hide"}
## OG Code, fixed by JBW below

# run pcadapt, test up to 20 clusters/pops
acp <- pcadapt(input = vcf, K = 5)

# Screeplot displays in decreasing order the percentage of variance explained by each PC
plot(acp, option = "screeplot",  K = 5)

#displays the PCA results colored by collection regions
pca <- plot(acp, option = "scores", gg.col=transp(cols, 1.0)) #+
 # geom_point(aes(color=poptab$host))

summary(acp)

# spit out % variance
acp$singular.values

# wrangle into a new dataframe
df1 <- data.frame(Species = poptab[,1], Loc = poptab[,2], ID = poptab[,3], pca = pca$data)
all_pca <- as.data.frame(df1)
cols <- c("C" = "#000000", "H" = "#999999", "P" = "#0072B2")
# back up shapes: 0,1,2,3,8

f <- ggplot(all_pca) + 
  geom_point(aes(all_pca$pca.PC_i, all_pca$pca.PC_j,
               color = all_pca$Species, 
               fill = all_pca$Species,
               shape = all_pca$Loc), 
               size = 4, 
               alpha = 0.85) +
  theme_bw() +
  theme(legend.position = "right",
        axis.title = element_text(face = "bold", size = 14),
        plot.title = element_text(face = "bold"),
        legend.title = element_text(face = "bold"),
        panel.grid = element_blank(),
        axis.text.x = element_blank(),
  axis.text.y = element_blank()) +
    scale_fill_manual(values = cols, 
                     breaks = c("C", "H", "P"),
                     labels = c("Acropora cervicornis", "Hybrid", "Acropora palmata"),
                     name = "S. 'fitti' host") +
  scale_color_manual(values = cols, 
                     breaks = c("C", "H", "P"),
                     labels = c("Acropora cervicornis", "Hybrid", "Acropora palmata"),
                     name = "S. 'fitti' host") +
    stat_ellipse(geom = "polygon", type = "t", alpha = 0.25, aes(all_pca$pca.PC_i, all_pca$pca.PC_j, fill= Species), show.legend = FALSE) + scale_fill_manual(values=cols) + # ellipses assumes multivariate distribution using default confidence level (0.95)
  scale_shape_manual(values = c(25,16,17,15,23), breaks = c("BA", "BE", "CU", "FL", "VI"), labels = c("Bahamas", "Belize", "Curacao", "Florida", "USVI"), name = "Location") +
  labs(title = "58,538 'genotyping' S. fitti SNPs", x="PC1 (16.0% variance)", y="PC2 (13.6% variance)")

f
```

```{r Jane's attempt to make code work}
# run pcadapt
acp <- pcadapt(input = vcf, K = 5)

# Screeplot
plot(acp, option = "screeplot",  K = 5)

summary(acp)

# spit out % variance... JBW gets PC1 = 15.7% and PC2 = 12.2%
acp$singular.values

# wrangle into a new dataframe >> JBW had to re-specify columns here (numbered cols above are the wrong ones)
df1 <- data.frame(Species = poptab$Host_Species, Loc = poptab$Population, 
                  ID = poptab$VCF_ID, pca = pca$data)
all_pca <- as.data.frame(df1)
cols2 = c("#000000","#999999","#0072B2") #JBW tweak to make colors work

f <- ggplot(all_pca, aes(x=pca.PC_i, y=pca.PC_j)) + #JBW moved x and y into general ggplot aesthetic
  geom_point(aes(color = Species, 
                 fill = Species,
                 shape = Loc),
             size = 4, 
             alpha = 0.85) +
  #JBW had to replace break values to actual place names, to match dataframe
  scale_shape_manual(values = c(25,16,17,15,23), 
                     breaks = c("Bahamas", "Belize", "Curacao", "Florida", "US Virgin Islands"), 
                     labels = c("Bahamas", "Belize", "Curacao", "Florida", "US Virgin Islands"), 
                     name = "Location") +
  scale_color_manual(values = cols2, 
                     breaks = c("A. cervicornis", "Hybrid", "A. palmata"),
                     labels = c("A. cervicornis", "Hybrid", "A. palmata"),
                     name = "S. 'fitti' host") +
  theme_classic2() +
  ylim(0.4, -0.4) + #JBW had to reverse the y-axis to get the spatial layout to match
  theme(legend.position = "right",
        axis.title = element_text(face = "bold", size = 14),
        plot.title = element_text(face = "bold"),
        legend.title = element_text(face = "bold"),
        panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_blank()) + 
  stat_ellipse(geom = "polygon", type = "t", alpha = 0.25, 
               aes(fill= Species), 
               show.legend = FALSE) + 
  #JBW combined 2 scale_fill_manual commands into 1; had to re-assign breaks to match names in dataframe
  scale_fill_manual(values = cols2, 
                    breaks = c("A. cervicornis", "Hybrid", "A. palmata"),
                    labels = c("A. cervicornis", "Hybrid", "A. palmata"),
                    name = "S. 'fitti' host") +
  # ellipses assumes multivariate distribution using default confidence level (0.95)
  labs(title = "57,810 'genotyping' S. fitti SNPs", x="PC1 (15.7% variance)", y="PC2 (12.2% variance)")
    #JBW edited plot title to match R output

f
```

```{r}
#JBW saved plot using code below (only run once)
#save_plot("Fig2_JBW.pdf", f, base_aspect_ratio = 1.6)
```