{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Making figure 5 (Manhattan Plot of outliers) from BayeScan and PCAdapt output\n",
    "```{r warning=FALSE, message=FALSE}\n",
    "# script modified from \"Garcia_AcervPopGenReport\" (From Dr. Sheila Kitchen)\n",
    "# load a zillion packages\n",
    "library(reshape2)\n",
    "library(pcadapt)\n",
    "library(qvalue)\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(boa)\n",
    "library(vcfR)\n",
    "library(gghighlight)\n",
    "library(qqman)\n",
    "library(cowplot)\n",
    "```\n",
    "\n",
    "\n",
    "```{r}\n",
    "# set working directory\n",
    "# import the vcf file\n",
    "vcf <- read.pcadapt(path, type = \"vcf\", type.out = \"matrix\", ploidy = 1)\n",
    "vcfr <- read.vcfR(\"vortexfs1/omics/env-bio/collaboration/dino_popgen/output/snp_calling/allhqSNPmm80.recode.vcf\")\n",
    "#pull out chrom and position from VCF file\n",
    "t <- getCHROM(vcfr)\n",
    "u <- getPOS(vcfr)\n",
    "r<- paste(t,u,sep=\"_\") #loci names\n",
    "c <- read.table(\"../output/seq_outliers/PCADAPT_ALLvalues.txt\", header=T)\n",
    "# read in the bayescan results\n",
    "b <- read.table(\"../output/seq_outliers/bayescan_outliers_clean.txt\", header=T)\n",
    "outliers<- c  %>%\n",
    "  mutate(PCA_outlier = ifelse(PCA.q.val < 0.05, \"TRUE\", \"FALSE\")) %>%\n",
    "  # mutate(PCA_pc1 = ifelse(out2$ID %in% snp_pc1$SNP, snp_pc1$PC, \"FALSE\")) %>%\n",
    "  left_join(b %>% select(\"BAYES_qval\",\"BAYES_fst\", \"ID\"), by=c(\"ID\")) %>%\n",
    "  mutate(BAYES_outlier = ifelse(ID %in% b$ID, \"TRUE\", \"FALSE\")) \n",
    "#separate(LocusName, into = c(\"chrom\", \"pos\"), sep=\"_(?=[^_]+$)\")\n",
    "# log -10 them\n",
    "outliers$PCA.q.val_LOG <- (log(outliers$PCA.q.val, base=10))*-1\n",
    "outliers$BAYES_qval_LOG <- (log(outliers$BAYES_qval, base=10))*-1\n",
    "bayes <- outliers %>%\n",
    "  group_by(BAYES_outlier) %>% \n",
    "  filter(BAYES_outlier == \"TRUE\") %>%\n",
    "  ungroup()\n",
    "pca <- outliers %>%\n",
    "  group_by(PCA_outlier) %>% \n",
    "  filter(PCA_outlier == \"TRUE\") %>%\n",
    "  ungroup()\n",
    "# add a fake chromosome column\n",
    "outliers$CHR <- c(\"1\")\n",
    "outliers$CHR <- as.numeric(outliers$CHR)\n",
    "bayes <- outliers %>%\n",
    "  group_by(BAYES_outlier) %>% \n",
    "  filter(BAYES_outlier == \"TRUE\") %>%\n",
    "  ungroup()\n",
    "bayes_h <- bayes$ID\n",
    "pca <- outliers %>%\n",
    "  group_by(PCA_outlier) %>% \n",
    "  filter(PCA_outlier == \"TRUE\") %>%\n",
    "  ungroup()\n",
    "pca_h <- pca$ID\n",
    "manhattan(subset(outliers, CHR == 1), bp = \"ID\", snp=\"ID\", p=\"PCA.p.val\", ylim=c(0,80), suggestiveline = FALSE, genomewideline = TRUE, highlight = c(bayes_h))  #Where to draw a \"genome-wide sigificant\" line. Default -log10(5e-8). \n",
    "# ggplot version https://www.r-graph-gallery.com/101_Manhattan_plot.html, https://cran.r-project.org/web/packages/qqman/vignettes/qqman.html\n",
    "# do this eventually https://github.com/sahirbhatnagar/manhattanly\n",
    "# remove inifinite values from Bayes Qvalue, editing figure caption to reflect this\n",
    "outliers[,5][outliers[,5] == 0] <- NA\n",
    "qv2 <- ggplot(outliers, aes(x=ID, y=-log10(PCA.q.val))) +\n",
    "  geom_point(subset(outliers, c(PCA.q.val_LOG <= 2.54)),mapping=aes(color = \"#999999\"), alpha=0.8, size=2) +\n",
    "  geom_point(data=subset(outliers, c(PCA_outlier==\"TRUE\" & PCA.q.val_LOG >= 2.54)), mapping=aes(color = \"blue\"), size = 0.9, alpha = 0.4) +\n",
    "  geom_point(data=subset(outliers, c(BAYES_outlier==\"TRUE\" & BAYES_qval_LOG >= 2.54)), aes(x=ID, y=-log10(BAYES_qval),color=\"red\"), size=0.9, alpha = 0.4) +\n",
    "  #  geom_point(data=subset(outliers, BAYES_outlier==\"TRUE\" & PCA_outlier==\"TRUE\"), mapping=aes(color=\"#D55E00\"), size=0.8) +\n",
    "  theme_bw() +\n",
    "  geom_hline(yintercept = 2.54,linetype = \"dashed\", lwd=0.6)+\n",
    "  geom_hline(yintercept = 6.01,lwd=0.6)+\n",
    "  labs(x=\"SNP ID #\", y=\"-log10(q-value)\") +\n",
    "  theme(legend.position = \"bottom\",\n",
    "        axis.text = element_text(size=12),\n",
    "        axis.title = element_text(size=14, face = \"bold\"),\n",
    "        legend.text = element_text(size=10),\n",
    "        legend.title = element_text(size=12, face = \"bold\")) +\n",
    "  scale_colour_manual(values = c(\"#999999\",\"blue\", \"red\"),\n",
    "                      labels = c(\"Not Significant\",\"PCAdapt\", \"BayeScan\"), name = c(\"Outlier Type\")) \n",
    "qv2\n",
    "qv1 <- ggplot(outliers, aes(x=ID, y=-log10(PCA.q.val))) +\n",
    "  geom_point(subset(outliers, c(PCA_outlier==\"FALSE\")),mapping=aes(color = \"#999999\"), alpha=0.8, size=2) +\n",
    "  geom_point(data=subset(outliers, c(PCA_outlier==\"TRUE\")), mapping=aes(color = \"blue\"), size = 0.9, alpha = 0.4) +\n",
    "  geom_point(data=subset(outliers, c(BAYES_outlier==\"TRUE\")), aes(x=ID, y=-log10(BAYES_qval),color=\"red\"), size=0.9, alpha = 0.4) +\n",
    "  #  geom_point(data=subset(outliers, BAYES_outlier==\"TRUE\" & PCA_outlier==\"TRUE\"), mapping=aes(color=\"#D55E00\"), size=0.8) +\n",
    "  theme_bw() +\n",
    "  geom_hline(yintercept = 2.54,linetype = \"dashed\", lwd=0.6)+\n",
    "  geom_hline(yintercept = 6.01,lwd=0.6)+\n",
    "  labs(x=\"SNP ID #\", y=\"-log10(q-value)\") +\n",
    "  theme(legend.position = \"bottom\",\n",
    "        axis.text = element_text(size=12),\n",
    "        axis.title = element_text(size=14, face = \"bold\"),\n",
    "        legend.text = element_text(size=10),\n",
    "        legend.title = element_text(size=12, face = \"bold\")) +\n",
    "  scale_colour_manual(values = c(\"#999999\",\"blue\", \"red\"),\n",
    "                      labels = c(\"Not Significant\",\"PCAdapt\", \"BayeScan\"), name = c(\"Outlier Type\")) \n",
    "qv1\n",
    "save_plot(\"../figures/Fig_5_manhattanplot_total.pdf\", qv1, base_aspect_ratio = 1.6)\n",
    "```\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
